### PR TITLE
[lldb][AIX] Header Parsing for XCOFF Object File in AIX

### PR DESCRIPTION
**Pull request by DhruvSrivastavaX**
This PR is in reference to porting LLDB on AIX.

Link to discussions on llvm discourse and github:

1. https://discourse.llvm.org/t/port-lldb-to-ibm-aix/80640
2. https://github.com/llvm/llvm-project/issues/101657
The complete changes for porting are present in this draft PR:
https://github.com/llvm/llvm-project/pull/102601

Added XCOFF Object File Header Parsing for AIX.
This PR is an incremental PR to the base:
- https://github.com/llvm/llvm-project/pull/111814

Details about XCOFF file format on AIX: [XCOFF](https://www.ibm.com/docs/en/aix/7.3?topic=formats-xcoff-object-file-format)

Review Request: @labath @DavidSpickett

### CODE REVIEW

**Comment by DhruvSrivastavaX** [comment]
Hi @labath , not sure why the `Labelling new pull requests` check failed. Also, the build check didnt automatically start for this one.

**Review by labath — COMMENTED** [review]
Do you actually need to parse the headers yourself? The reason for using the llvm implementation was so that we can avoid that. FWICS, all of this data should be available through the `(file|auxiliary)Header(32|64)` APIs.

BTW, this is exactly the kind of granularity I was looking for.

**Comment by labath** [comment]
> Hi @labath , not sure why the `Labelling new pull requests` check failed. Also, the build check didnt automatically start for this one.

I don't know, but I wouldn't be worried about that. The test failure is more of a concern.

**Comment by DhruvSrivastavaX** [comment]
> Do you actually need to parse the headers yourself? The reason for using the llvm implementation was so that we can avoid that. FWICS, all of this data should be available through the `(file|auxiliary)Header(32|64)` APIs.

Okay, I'll take a look at that. 



> BTW, this is exactly the kind of granularity I was looking for.

Happy to hear that! Will follow the similar practice.

> I don't know, but I wouldn't be worried about that. The test failure is more of a concern.

Yeah, I think I missed the test case modification accordingly. Will update!

**Comment by DhruvSrivastavaX** [comment]
It is updated now. Please provide your review and merge if everything is good. 
@labath

**Comment by labath** [comment]
So, using the llvm structures for parsing (instead of rolling your own) is definitely a step in the right direction. However, I don't think you need to roll your own parsing either, given that the headers have already been parsed in llvm::XCOFFObjectFile. Why can't you just use the header from there?

**Comment by DhruvSrivastavaX** [comment]
Yes, I did try a few things while rewriting but i was seeing some stray values which were not actually present in the raw file. 
So just want to make sure that I am using it correctly.

File: lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp
Comment (labath) [inline lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp:98]:
```suggestion
  auto binary = llvm::object::XCOFFObjectFile::create(
```

And then you don't need the dyn_cast below. (You do need to make the create function public, but I don't think that should be controversial given that it's already public for all other object file types)

File: lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp
Comment (labath) [inline lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp:177]: Given that this is called directly from CreateInstance, I'm pretty sure this locking is not needed (ObjectFileELF definitely doesn't do that, although PECOFF for some reason does).

File: lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp
Comment (labath) [inline lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp:203]: If you're going to be supporting only the 64-bit version, then I think the previous version is actually better. :)

File: lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.h
Comment (labath) [inline lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.h:113]: Can we get rid of these variables? You can access the headers directly from the m_binary member, right?

Comment (DhruvSrivastavaX) [reply lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.h:113]: The reason to keep these is for the ease of access and avoid calling the functions like `(file|auxiliary)Header(32|64)` functions repetitively whenever the values are required in different functions.

File: lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp
Comment (DhruvSrivastavaX) [reply lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp:203]: Yes sure. 32 bit support is anyway a future item.

File: lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.h
Comment (labath) [reply lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.h:113]: What's the issue with calling those functions? It can't be performance, as these functions merely return a pointer the the member inside the llvm object file. If you find too `m_binary->auxiliaryHeader64()` long/repetitive, I'd suggest wrapping that in an accessor functions -- though that will only save a few chars, so I'm not sure if its worth it (it may be better to have an accessor for the fields of the header that are used often -- that would also be a way to abstract the 64/32 difference). And if you're worried about the `m_binary` pointer being null, then I'd suggest setting up the code in a way that its non-nullness is guaranteed (for example, by creating the XCOFFObjectFile in the CreateInstance function and passing it to the constructor.

Bottom line: I still believe there's no need for this member and there's a better way to achieve what you're trying to do. I'm not sure what it is, because I'm don't know what exactly you are trying to achieve.

Comment (DhruvSrivastavaX) [reply lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.h:113]: Sure then. I will replace it thats not a problem.

File: lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp
Comment (DhruvSrivastavaX) [reply lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp:98]: One concern here is, there are two instances of this: `create` and `createXCOFFObjectFile` both of which are private. 
Also they expect a `TYPE_ID` as a paramater, which is again a protected member of the `Binary` class. 
`createObjectFile` is the only accessible function that accepts a public `file_magic::xcoff_object_64` type of parameter and internally calls the appropriate `create(XCOFF|ELF|others)ObjectFile` APIs and passes thec correct TYPE_ID too.
How shall we deal with it in this scenario?

File: lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp
Comment (DhruvSrivastavaX) [reply lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp:177]: Done

File: lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp
Comment (DavidSpickett) [inline lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp:182]: Would be simpler to just `return true`. Or:
```
// Only 64-bit is supported for now
return module_sp && m_binary->fileHeader64()->Magic == XCOFF::XCOFF64;
```
For extreme minimalism :)

File: lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp
Comment (DavidSpickett) [inline lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp:195]: https://llvm.org/docs/CodingStandards.html#comment-formatting - use `//`. If you want a block just use multiple `//` lines.

Comment (DavidSpickett) [reply lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp:195]: Also I'm a bit wary of adding more TODOs that we all forget to get to :) . Perhaps just `// 32 bit is not supported.`.

Then if you get time to do the work to support it, great, if you don't, also fine.

File: lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.h
Comment (DavidSpickett) [inline lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.h:20]: As before, if clang-format is making these changes, make a separate PR for this and we can merge it swiftly.

File: lldb/test/Shell/ObjectFile/XCOFF/basic-info.yaml
Comment (DavidSpickett) [inline lldb/test/Shell/ObjectFile/XCOFF/basic-info.yaml:16]: Not sure if we're allowed comments in this YAML. But could you at least confirm to me here (in reply to this comment I mean), what 2 means? Is this the EXEC flag?

I guess that this test started failing because you actually made the parsing better. Is that the case?

**Review by DavidSpickett — COMMENTED** [review]
No big objections just small things.

If you make a clang-format PR add me as a reviewer, I'll get it merged quickly so you can rebase this.

Seems like the sticking point is create vs. createXCOFFObjectFile and I'll leave that to you and Pavel to sort out.

File: lldb/test/Shell/ObjectFile/XCOFF/basic-info.yaml
Comment (DhruvSrivastavaX) [reply lldb/test/Shell/ObjectFile/XCOFF/basic-info.yaml:16]: Yes, that is the exact reason for this change. A 2 here implies that it is an exec.

File: lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.h
Comment (DhruvSrivastavaX) [reply lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.h:20]: Ok great! By these you mean just the includes and smaller formatting fixes? Let me drop those right away!

File: lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp
Comment (DhruvSrivastavaX) [reply lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp:195]: Sure, that should be better.

File: lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp
Comment (DhruvSrivastavaX) [reply lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp:98]: Addressed all the other comments. Please do let me know which one to proceed with, as per the above comment.
Thanks!

Comment (DhruvSrivastavaX) [reply lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp:98]: Yes, that will be good. In that case, I can open a PR under llvm code to split the createXCOFFObjectfile to `32|64`. 
If it goes through I can introduce the change for objectfilexcoff too, if we are able to introduce. 
For this one, I can change try changing to `llvm::object::ObjectFile::createObjectFile`

Comment (DhruvSrivastavaX) [reply lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp]: Corrected it here @labath.

File: lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp
Comment (labath) [inline lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp:110]: Remove blank line (and reformat)

File: lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp
Comment (labath) [inline lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp:175]:
```suggestion
  return m_binary->fileHeader64()->Magic == XCOFF::XCOFF64;
```

**Review by labath — APPROVED** [review]
Two more quick comments, but otherwise looks good. Thanks for your patience.

File: lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp
Comment (labath) [reply lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp:175]: (you can keep the comment if you want. My point is you don't need to check the module validity)

File: lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp
Comment (DhruvSrivastavaX) [reply lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp:109]: Done

File: lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp
Comment (DhruvSrivastavaX) [reply lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp]: Sure, Done

**Comment by DhruvSrivastavaX** [comment]
So, I have handled all the comments. Please merge if everything is alright. 
@labath @DavidSpickett

**Review by DavidSpickett — APPROVED** [review]
LGTM

**Comment by DhruvSrivastavaX** [comment]
So, all the tests have passed. 
I would request you to merge as I dont have the write access yet. 
Thanks!!

**Pull request merged**
